### ISSUE TITLE
sd3.01  Automatic Metadata Git Push - some questions

### ISSUE THREAD

**Issue by rgleason** [Rasbats/shipdriver_pi#389]
Is this all that is necessary  to push to Master branch and Cloudsmith <plugin>-prod?

Assuming this has been executed from the master branch:
```
git commit -am "1.2.3  release to master branch"
git push origin master  
```
  (which should go to the Cloudsmith  <plugin>-prod repository!   -- but this goes to <plugin>-alpha now!! Why???

Then we do [Syncing Branches](https://opencpn-manuals.github.io/main/AlternativeWorkflow/InstallConfigure/Catalog-Github-Integration.html#_syncing_branches) as above.  (not using the last two beta commands)

How does the plugins/auto branch get updated to match upstream/master?  Is this automatic somehow?

Then create the tag and push it
```
git tag v1.2.3
git push origin v1.2.3
```

Will this then cause Circleci to deploy the plugin files to [CS <plugin>-prod](https://cloudsmith.io/~opencpn/repos/logbookkonni-prod/packages/)  and then to [execute the "git-push.sh"](https://app.circleci.com/pipelines/github/rgleason/LogbookKonni_pi/138/workflows/572f5816-2925-497b-8511-23222369fb60/jobs/1477) and copy the metadata files to github.com/rgleason/plugins/auto  for example?

**Comment by leamas** [comment]
> Will this then cause Circleci to deploy the plugin file [...]

No. Just pushing the tag will not affect anything, what counts is the tags existing when making the build. Two ways to cope with it.
  - The first is to rebuild the same commit again after pushing the tag.  This corresponds to a workflow where you first make a  build, test it and eventually tag it and rebuild. Common in larger, professionally handled software projects.
  - The second is to push the tag and the branch in one step, corresponding to a workflow where you don't really test the last commit. Done using something like ` git push origin --atomic refs/tags/v1,2.3:refs/tags/v1.2.3  master:master` assuming a tag v1.2.3 and working with master branch

**Comment by rgleason** [comment]
Ok, thanks very much, I had forgotten about that dialog with Hakan, so do it in this order?

1.  Update your github  **_plugins/master_**  [Syncing Branches](https://opencpn-manuals.github.io/main/AlternativeWorkflow/InstallConfigure/Catalog-Github-Integration.html#_syncing_branches) as above.  (not using the last two beta commands)

_How does the **_plugins/auto branch_** get updated to match _**upstream/master**_?  Is this automatic somehow?_

2.  From your plugin,   initiate a tagged push from your plugin  and then create an empty commit.

   git tag v1.2.3
   git push v1.2.3
   git commit --allow-empty -m "Trigger build"
   git push origin 

OR do it all at the same time

   git tag v1.2.3
   git commit -am "1.2.3  release to master branch"
   git push origin --atomic refs/tags/v1.2.3:refs/tags/v1.2.3 master:master

**Comment by rgleason** [comment]
Doing this first step
1.  Update your github  **_plugins/master_**  [Syncing Branches](https://opencpn-manuals.github.io/main/AlternativeWorkflow/InstallConfigure/Catalog-Github-Integration.html#_syncing_branches) as above.  (not using the last two beta commands)

_How does the **_plugins/auto branch_** get updated to match _**upstream/master**_?  Is this automatic somehow?_

....or do I have to get that current with the plugins/master branch every time before pushing a tagged commit?

**Comment by rgleason** [comment]
Since I had already pushed the tag, I then
```
   git commit --allow-empty -m "Trigger build"
   git push origin 
```

and many of the files built and were pushed to  logbookkonni-alpha instead of "prod"
also the files "failed" because they were not pushed to plugins/auto.

I have no idea if I have this configured properly because  CS is still misbehaving.

**Comment by rgleason** [comment]
`./update-templates -l
`
Adds a new remote!!   This is necessary to avoid the previous errors.
The manual should mention this as a first necessary step.

```
fcgle@Hesper MINGW64 ~/source/logbookkonni_pi (master)
$ update-templates -l
bash: update-templates: command not found

fcgle@Hesper MINGW64 ~/source/logbookkonni_pi (master)
$ ./update-templates -l
Adding new shipdriver remote
sd3.0.0
sd3.0.0-beta1
sd3.0.0-beta2
sd3.0.0-rc1
sd3.0.1
sd3.0.1-beta1
sd3.0.1-beta4
sd3.0.1-beta5
sd3.0.2-beta1

fcgle@Hesper MINGW64 ~/source/logbookkonni_pi (master)
$ git remote -v
origin  git@github.com:rgleason/LogbookKonni_pi.git (fetch)
origin  git@github.com:rgleason/LogbookKonni_pi.git (push)
shipdriver      https://github.com/Rasbats/shipdriver_pi.git (fetch)
shipdriver      https://github.com/Rasbats/shipdriver_pi.git (push)
upstream        https://github.com/delatbabel/LogbookKonni-1.2.git (fetch)
upstream        https://github.com/delatbabel/LogbookKonni-1.2.git (push)


```

**Comment by rgleason** [comment]
I just received  an appveyor notice of success!  git-push worked.
https://ci.appveyor.com/project/rgleason/logbookkonni-pi/builds/42131504

One lone     Logbook-1.4-msvc-10.0.14393.xml   [ here](https://github.com/rgleason/plugins/tree/auto/metadata)

**Comment by leamas** [comment]
But last time, you still struggled  to get git-push to work also in circleci.  At the time, I could not understand the situation in any other way than that you had the GIT_KEY_PASSWORD correct in appveyour but not in circleci.

Also, it seems that you created a new key with a new password. This might have been overkill. There actually is a tool _ci/check-key-password_ for situations like this. Using this, you can verify the password for a key file.  If you are sure about the password, it's just to re-enter it into CircleCI. 

Common source of errors is copy-pasting extra space before or after the pw.

**Comment by rgleason** [comment]
Yes, I've checked for spaces.  I will use ci/check-key-password  to get this right. Thanks.

**Comment by rgleason** [comment]
fcgle@Hesper MINGW64 ~/source/logbookkonni_pi (v3.0)
$ python --version
Python 3.10.0

fcgle@Hesper MINGW64 ~/source/logbookkonni_pi (v3.0)
$ ./check-key-password
bash: ./check-key-password: No such file or directory

(Doesn't this mean that python did not find the file "check-key-password"?

also tried 

fcgle@Hesper MINGW64 ~/source/logbookkonni_pi (v3.0)
$ ./ci/check-key-password
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

**Comment by leamas** [comment]
Assuming git-bash or linux. 

The file lives in _ci/check-key-password_.  Typical invocation is then 

    $ ci/check-key-password build-deps/username.enc <my secret password>

If this does not work your python installation is broken. In my windows case it is installed in _/C/Python310_. This is git-bash parlance, you need to check  your situation. Add that corresponding path element to the beginning of $PATH/%PATH%. In my case and using git-bash:`export PATH=/C/Python310:$PATH`

**Comment by rgleason** [comment]
That worked!    Now have the correct PW. Thanks!!

**Comment by rgleason** [comment]
I should have looked at the script instructions. Sorry.

**Comment by rgleason** [comment]
I am happy to report that it has built and pushed to plugins/auto  all OS except  (2)android and (2)raspbian
all in jig time.   MacOS is just 6 minutes.  Awesome.

**Comment by leamas** [comment]
The flatpak builds is something that differs from my builds. I will look into it.

The Android builds fails because the sources has to be updated to build also on Android. It has nothing to do with the templates

**Comment by rgleason** [comment]
Yes, Thanks for all the help.

**Comment by leamas** [comment]
I have made a new build. The diff against your v3.0 is basically empty

```
$ $ git rev-parse rgleason/v3.0
fd46a2758902943a649193c94b5bed7e7118135f

$ git diff  rgleason/v3.0 
diff --git a/Plugin.cmake b/Plugin.cmake
index 2d2ded9..d861aad 100644
--- a/Plugin.cmake
+++ b/Plugin.cmake
@@ -35,7 +35,7 @@ option(PLUGIN_USE_SVG "Use SVG graphics" ON)
 #  Using Shipdriver version 3.0.0 from branch v3.0
 
 set(PKG_NAME LogbookKonni_pi)
-set(PKG_VERSION  1.4.26.0)
+set(PKG_VERSION  1.4.25.0)
 set(PKG_PRERELEASE "")  # Empty, or a tag like 'beta'
 
 set(DISPLAY_NAME Logbook)    # Dialogs, installer artifacts, ...
diff --git a/cmake/TemplateVersion b/cmake/TemplateVersion
index ffb9142..798ecb5 100644
--- a/cmake/TemplateVersion
+++ b/cmake/TemplateVersion
@@ -1,4 +1,4 @@
 # Created by update-templates
-date: 2022-01-09 00:53 UTC
+date: 2022-01-10 09:48 UTC
 commit: e49825d
 tags: sd3.0.2-beta1
```

For me, everything builds except flatpak: [logs](https://app.circleci.com/pipelines/github/leamas/LogbookKonni_pi/67/workflows/226017ea-accb-466d-a4b6-b7bbf3b60c44)

So, why do the same sources build for me but not for you?  "puzzled"

**Comment by leamas** [comment]
I made a new build with sources which are identical with rgleason/v3.0 which resolves to fd46a27, the same commit which you built [above](https://github.com/Rasbats/shipdriver_pi/issues/389#issuecomment-1009002988). 

The logs are [here](https://app.circleci.com/pipelines/github/leamas/LogbookKonni_pi/68/workflows/66a1ffe0-0ba5-47df-ab71-ba33ed5533d8). Everything builds besides Android. 

What happens if you rebuild the one you linked to  at fd46a27?

**Comment by rgleason** [comment]
I've resrun from when stopped and they still fail with 

```
#!/bin/bash -eo pipefail
cd /build-flatpak; /bin/bash < upload.sh

Using CLOUDSMITH_API_KEY: c148...
+ cloudsmith push raw --no-wait-for-sync --name Logbook-1.4-flatpak-A64-20.08-metadata --version 1.4.26.0+1550.fd46a27 --summary 'Plugin metadata for automatic installation' --republish opencpn/logbookkonni-alpha Logbook-1.4-flatpak-A64-20.08.xml
Usage: cloudsmith push raw [OPTIONS] OWNER/REPO
                           PACKAGE_FILE
Try 'cloudsmith push raw -h' for help.

Error: Invalid value for 'PACKAGE_FILE': File 'Logbook-1.4-flatpak-A64-20.08.xml' does not exist.

Exited with code exit status 2

CircleCI received exit code 2

```

Did I miss some permissions?

**Comment by rgleason** [comment]
I have merged these changes to my local (only) master branch, but have not pushed yet.  Maybe I'll stay on v3.0 branch until this gets sorted out.

**Comment by rgleason** [comment]
Is it not building  Logbook-1.4-flatpak-A64-20.08.xml   for some reason?

**Comment by leamas** [comment]
It fails here in the build phase:
```
Initialized empty Git repository in /home/circleci/project/build-flatpak/.flatpak-builder/git/file_home_circleci_project-SVKSF1/
Fetching git repo file:///home/circleci/project, ref refs/heads/master
From file:///home/circleci/project
 * [new branch]      master     -> master
Failed to download sources: module Logbook: Not a gitlink tree: opencpn-libs
```

So, something about opencpn-libs.. You need to get rid of it completely:

  - in .git/config
  - in .git/modules
  - remove the directory opencpn-libs if it exists
  - remove .gitmodules (double check...]

**Comment by leamas** [comment]
> s it not building Logbook-1.4-flatpak-A64-20.08.xml for some reason?

It does not build anything at all, it cannot even clone the sources. See above.

**Comment by rgleason** [comment]
in .git/config   - no reference to opencpn-libs in here
    in .git/modules   - don't have this folder / file
    remove the directory opencpn-libs if it exists    - removed
    remove .gitmodules (double check...]    - removed previously, don't see it here now;

**Comment by leamas** [comment]
Well, then. try to rebuild. If it fails, look in the logs for the build step. You should see how it clones the sources in the beginning (compare to my builds which works.).

**Comment by rgleason** [comment]
I find in my remote   https://github.com/rgleason/LogbookKonni_pi/blob/master/.gitmodules
and I don't find opencpn-libs folder,  Perhaps I had not initialized it or registered it with git or something?

**Comment by rgleason** [comment]
Ok now that is the problem,  .gitmodules does not exist locally, but it does exist in my remote.
How to remove it?  Should I just remove it using tools in my remote to delete it and make a new commit?
Then fetch that change to local?

.git on-line looks ok non of the things you mentioned...

**Comment by leamas** [comment]
You remove in your local branch, and then push that change.

**Comment by leamas** [comment]
> Ok now that is the problem, .gitmodules does not exist locally, but it does exist in my remote.

This makes no sense. The remote as such does not contain anything. What does include things is remote branches, the branches on github. You update them by pushing to them. The remote is just a handle to access these branches.

This is a bit of sidetracking. What's in the master branch is irrelevant since you build the v3.0 branch.  OTOH, cleaning up master does not hurt, but does it help?

**Comment by leamas** [comment]
A look in circleci logs reveals that you tried to rebuild the master branch which failed miserably. Put this aside for now, not interesting.

Instead, working locally, go the directory above LogBookKonni_pi. In that, try to make a local clone to see if your repo indeed is broken

    $ git clone -b v3.0 LogbookKonni_pi LogbookKonni_pi-new

If it works, it creates a clean copy of your repo. If not, you can at least reproduce the problem locally. 

If it works, go to the new directory, add the remote and push from there to see if it helps:

    $ cd LogbookKonni_pi-new
    $ git remote remove origin
    $ git remote add origin git@github.com:rgleason/LogbookKonni_pi.git
    $ git commit --allow-empty -m "Trigger build"
    $ git push origin v3.0:v3.0

**Comment by leamas** [comment]
Good night.

**Comment by rgleason** [comment]
Sorry Alec I was out sawing plywood for window openings.

```
fcgle@Hesper MINGW64 ~/source

$ git clone -b v3.0 LogbookKonni_pi LogbookKonni_pi-new
Cloning into 'LogbookKonni_pi-new'...
done.

```

**Comment by rgleason** [comment]
I've merged v3.0 and pushed it to remote and now it is building flatpak.
